test/test_type_promotion.py

```diff
@@ -138,39 +139,39 @@ def test_mixed_type_backward(self):
         f_dtypes = [torch.float, torch.double]
         f_dtypes = f_dtypes if self.device == 'cpu' else f_dtypes + [torch.half]
         i_dtypes = [torch.int, torch.long]
-        for f in ['add', 'sub', 'rsub', 'mul', 'div']:
+        for func in [torch.add, torch.sub, torch.rsub, torch.mul, torch.div]:
             for dtype1 in f_dtypes:
                 for dtype2 in (f_dtypes + i_dtypes):
                     x = torch.ones(10, requires_grad=True, dtype=dtype1, device=self.device)
                     y = torch.ones(10, dtype=dtype2, device=self.device)
-
-                    func = getattr(torch, f)
                     func(x, y).sum().backward()
 
-    # verifies that a.add(b) is the same as a.to(b.dtype).add(b) in cases
-    # where that should hold.
+    # verifies that torch.<op>(a, b) is the same as 
+    # torch.<op>(a.to(common_dtype), b.to(common_dtype)) in cases where that should hold.
     def test_many_promotions(self):
-        from_to = {
-            torch.float16: torch.float32,
-            torch.half: torch.float16,
-            torch.int: torch.long,
-            torch.uint8: torch.long,
-            torch.uint8: torch.float,
-            torch.int: torch.float,
-            torch.int: torch.double,
-            torch.int16: torch.long,
-            torch.float16: torch.double,
-            torch.bool: torch.long,
-            torch.bool: torch.float
-        }
-
-        for k, v in from_to.items():
-            a = torch.rand([3, 3], device=self.device).to(k)  # no _th_uniform for half on cpu.
-            b = torch.rand([3, 3], device=self.device).to(v)
-            c = a.add(b)
-            d = a.to(v).add(b)
-            self.assertEqual(c.dtype, d.dtype, message='from {} to {}'.format(k, v))
-            self.assertEqual(c, d, message='from {} to {}'.format(k, v))
+        all_dtypes = torch.testing.get_all_math_dtypes(self.device)
+        for dt1, dt2 in itertools.product(all_dtypes, all_dtypes):
+            for op in [torch.add, torch.sub, torch.mul, torch.div]:
+                for contiguous in [True, False]:
+                    common_dtype = torch.promote_types(dt1, dt2)
+                    if op == torch.sub and common_dtype != torch.bool:
+                        # Subtraction, the `-` operator, with a bool tensor is not supported.
+                        continue
+
+                    first = torch.rand([20, 20], device=self.device).to(dt1)  # no _th_uniform for half on cpu.
+                    second = torch.rand([20, 20], device=self.device).to(dt2)
+                    # test ops with non-contiguous tensors
+                    if not contiguous:
+                        first = first.t()
+                        second = second.t()
```

Comment:
It might be good to create two non-contiguous tensors that are strided differently. Sometimes code fails on some non-contiguous inputs but doesn't fail if the non-contiguous tensors are strided in the same way.

Reply:
changed to make second be non-contiguous through indexing, but that means I need to be a bit funny about how I create the initial tensor. let me know if that's sufficiently readable to you.

Reply:
Couldn't we just create something like 2 three-dimensional tensors and then transpose different dimensions? The indexing is indeed difficult to read

Reply:
that works thanks, updated